### PR TITLE
GH-33807: [R] Add a message if we detect running under emulation

### LINKED ISSUE #33807: [R] Using dplyr::tally with an Arrow FileSystemDataset crashes R
### Describe the bug, including details regarding any error messages, version, and platform.

The following code snippet crashes R. I'm using arrow 10.0.1

```
library(dplyr)
arrow::write_dataset(cars, here::here("cars.feather"), format = "feather")
a <- arrow::open_dataset(here::here("cars.feather"), format = "feather")
a %>% tally()
```

**Platform information**
```
> sessionInfo()
R version 4.2.2 (2022-10-31)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Monterey 12.6

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] arrow_10.0.1   testthat_3.1.6

loaded via a namespace (and not attached):
 [1] assertthat_0.2.1 brio_1.1.3       R6_2.5.1         lifecycle_1.0.3  magrittr_2.0.3   rlang_1.0.6     
 [7] cli_3.5.0        rstudioapi_0.14  vctrs_0.5.1      tools_4.2.2      bit64_4.0.5      glue_1.6.2      
[13] purrr_1.0.0      bit_4.0.5        compiler_4.2.2   tidyselect_1.2.0
```

### Component(s)

R

### LINKED ISSUE #37760: existing_data_behavior = "overwrite" does not work as expected in   arrow::write_dataset
### Describe the bug, including details regarding any error messages, version, and platform.

It seems that the default for existing_data_behavior='overwrite' does not actually overwrite existing data. @jonkeane 

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

path <- file.path(tempdir(), "mtcars-data")

mtcars %>% 
  group_by(cyl) %>% 
  arrow::write_dataset(path)

list.dirs(path, full.names = F)
#> [1] ""      "cyl=4" "cyl=6" "cyl=8"
fs::file_size(path)
#> 160
  
mtcars %>% 
  group_by(vs) %>% 
  arrow::write_dataset(path, existing_data_behavior = "overwrite")

list.dirs(path, full.names = F)
#> [1] ""      "cyl=4" "cyl=6" "cyl=8" "vs=0"  "vs=1"
fs::file_size(path)
#> 224

sessionInfo()
#> R version 4.3.1 (2023-06-16)
#> Platform: aarch64-apple-darwin20 (64-bit)
#> Running under: macOS Ventura 13.5.2
#> 
#> Matrix products: default
#> BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
#> LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0
#> 
#> locale:
#> [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#> 
#> time zone: America/Chicago
#> tzcode source: internal
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] dplyr_1.1.3
#> 
#> loaded via a namespace (and not attached):
#>  [1] vctrs_0.6.3       cli_3.6.1         knitr_1.44        rlang_1.1.1      
#>  [5] xfun_0.40         purrr_1.0.2       generics_0.1.3    assertthat_0.2.1 
#>  [9] bit_4.0.5         glue_1.6.2        htmltools_0.5.6   fansi_1.0.4      
#> [13] rmarkdown_2.24    evaluate_0.21     tibble_3.2.1      fastmap_1.1.1    
#> [17] yaml_2.3.7        lifecycle_1.0.3   compiler_4.3.1    fs_1.6.3         
#> [21] pkgconfig_2.0.3   rstudioapi_0.15.0 digest_0.6.33     R6_2.5.1         
#> [25] reprex_2.0.2      tidyselect_1.2.0  utf8_1.2.3        pillar_1.9.0     
#> [29] magrittr_2.0.3    bit64_4.0.5       tools_4.3.1       withr_2.5.0      
#> [33] arrow_13.0.0
```

<sup>Created on 2023-09-17 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### Component(s)

R

### PR DESCRIPTION
**Pull request by jonkeane**
Resolves #33807 and #37034

### Rationale for this change

If someone is running R under emulation, arrow segfaults without error. We can detect this when we load so can also warn people that this is not recommended. Though the version of R being run is not directly an arrow issue, arrow fails very quickly in this configuration.

### What changes are included in this PR?

Detect when running under rosetta (on macOS only) and warn when the library is attached

### Are these changes tested?

No, given the paucity of ARM-based mac CI, testing this organically would be difficult. But the logic is straightforward.

### Are there any user-facing changes?

Yes, a warning when someone loads arrow under emulation.
* Closes: #33807

### CODE REVIEW

File: r/R/arrow-package.R [hunk 1]
```diff
@@ -196,6 +196,20 @@ configure_tzdb <- function() {
         )
       )
     }
+
+    # On MacOS only, Check if we are running in under emulation, and warn this will not work
+    rosetta <- identical(sysname, "darwin") && identical(system("sysctl -n sysctl.proc_translated", intern = TRUE), "1")
+    if (rosetta) {
+      packageStartupMessage(
```
Comment (jonkeane) [inline r/R/arrow-package.R:203]: Should this actually be a `warning()`? Or even a `stop()`?

File: r/R/arrow-package.R [hunk 2]
```diff
@@ -196,6 +196,20 @@ configure_tzdb <- function() {
...
+        paste(
+          "It appears that you are running R and Arrow in emulation",
+          "(typically, this is running an x86 build of R on an ARM or",
+          "M processor based mac). This configuration is not supported",
+          "by arrow and we highly recommend you install a native (arm64)",
+          "build of R and use arrow with that."
```
Comment (jonkeane) [inline r/R/arrow-package.R:209]: This is more verbose than many of our messages (and I'm happy for some streamlining), but my reasoning for being _so_ explicit is that if someone finds themselves in this situation, they might not even realized what is going on with architectures, versions, emulation etc. So I wanted to give people more details to be able to get to a good state (native R + native arrow)

File: r/R/arrow-package.R (hunk 1 above)
Comment (thisisnic) [reply r/R/arrow-package.R:203]: I guess it depends on the extent of the problem; does it always segfault, or just when we use some of the functionality?

Comment (jonkeane) [reply r/R/arrow-package.R:203]: There might be some functionality that works, but everything I've seen people try (even `open_dataset() |> nrow()` segfaults. Given opcodes + architectures + emulation, I would be surprised if much of anything of use in arrow will work.

File: r/R/arrow-package.R (hunk 2 above)
Comment (nealrichardson) [reply r/R/arrow-package.R:209]: I might explicitly link to https://cran.r-project.org/bin/macosx/ and say install the arm64 version. And I (personally) would be a bit more direct: you're running R built for Intel on a non-Intel mac, and arrow won't work in that context; to fix, install the arm64 version of R from CRAN and then reinstall R.

File: r/R/arrow-package.R (hunk 1 above)
Comment (nealrichardson) [reply r/R/arrow-package.R:203]: I think `packageStartupMessage` is appropriate, I wouldn't raise other signals in onLoad

File: r/R/arrow-package.R [hunk 3]
```diff
@@ -196,6 +196,20 @@ configure_tzdb <- function() {
         )
       )
     }
+
+    # On MacOS only, Check if we are running in under emulation, and warn this will not work
+    rosetta <- identical(sysname, "darwin") && identical(system("sysctl -n sysctl.proc_translated", intern = TRUE), "1")
```
Comment (nealrichardson) [inline r/R/arrow-package.R:201]: I see this is also in install_arrow(), NBD but we could factor this out and put it in arrow-package.R alongside `on_macos_10_13_or_lower()` et al.

**Comment by nealrichardson** [comment]
> ## Are these changes tested?
>
> No, given the paucity of ARM-based mac CI, testing this organically would be difficult. But the logic is straightforward.

Have you confirmed locally at least?

**Comment by nealrichardson** [comment]
Also resolves #37034?

**Comment by nealrichardson** [comment]
Should this get mentioned in the docs somewhere too?

**Comment by jonkeane** [comment]
> Have you confirmed locally at least?

Yes, here is an example of the output. After seeing it in the terminal I've moved it up about the missing features message + also add "Warning:" and some spacing to hopefully set this apart a little bit more.

```
Platform: x86_64-apple-darwin17.0 (64-bit)
...
> library(arrow)
Warning:
  It appears that you are running R and Arrow in emulation (i.e. you're
  running an Intel version of R on a non-Intel mac). This configuration is
  not supported by arrow, you should install a native (arm64) build of R
  and use arrow with that. See https://cran.r-project.org/bin/macosx/

Some features are not enabled in this build of Arrow. Run `arrow_info()` for more information.

Attaching package: ‘arrow’

The following object is masked from ‘package:utils’:

    timestamp
```

**Comment by jonkeane** [comment]
> Should this get mentioned in the docs somewhere too?

Add this to the readme section under caveats.

File: r/R/arrow-package.R (hunk 2 above)
Comment (jonkeane) [reply r/R/arrow-package.R]: Thanks for the "Intel on a non-Intel mac" that's probably the clearest way to describe it to folks who are most likely to run into this.

**Review by nealrichardson — APPROVED** [review]
Thanks for doing this!

**Pull request merged**